src/Disks/IDiskTransaction.h

```diff
@@ -131,6 +135,9 @@ struct IDiskTransaction : private boost::noncopyable
 
     /// Truncate file to the target size.
     virtual void truncateFile(const std::string & src_path, size_t target_size) = 0;
+
+protected:
+    UInt64 remove_shared_recursive_batch_size;
```

Comment:
Why should it be the parameter for `IDiskTransaction`? Are there any blockers from keep it in `RemoveRecursiveObjectStorageOperation` only?

In that case we can pass the batch size like: 
```
void DiskObjectStorageTransaction::removeSharedRecursive(
    const std::string & path, bool keep_all_shared_data, const NameSet & file_names_remove_metadata_only), size_t batch_size)

```

Reply:
I tried to do that, but there is a place in [code](https://github.com/ClickHouse/ClickHouse/blob/a05312c6541d4c761fd5c242dba372d2cff5e842/src/Storages/MergeTree/DataPartStorageOnDiskBase.h#L130) ([usage](https://github.com/ClickHouse/ClickHouse/blob/e18f2448777bbe2128ade9f99ec54fdec949f23d/src/Storages/MergeTree/DataPartStorageOnDiskBase.cpp#L646)) that assumes that `removeSharedRecursive`'s(and other methods) signature in `IDisk` and `IDiskTransaction` is the same, that's why I decided it will be simpler to add a field to `IDiskTransaction` instead

Reply:
It looks like we can keep it in `DiskObjectStorage`. Mentioned code calls disk method, not transaction.

Reply:
Building with `size_t batch_size` in `IDiskTransaction` signature throws this
```
/home/kirillgarbar/ClickHouse/src/Storages/MergeTree/DataPartStorageOnDiskBase.cpp:621:93: error: too few arguments to function call, expected 4, have 3
  621 |             executeWriteOperation([&](auto & disk) { disk.removeSharedRecursive(to, true, {}); });
      |                                                      ~~~~~~~~~~~~~~~~~~~~~~~~~~             ^
/home/kirillgarbar/ClickHouse/src/Storages/MergeTree/DataPartStorageOnDiskBase.h:133:13: note: in instantiation of function template specialization 'DB::DataPartStorageOnDiskBase::rename(std::string, std::string, LoggerPtr, bool, bool)::(anonymous class)::operator()<DB::IDiskTransaction>' requested here
  133 |             op(*transaction);
      |             ^
/home/kirillgarbar/ClickHouse/src/Storages/MergeTree/DataPartStorageOnDiskBase.cpp:621:13: note: in instantiation of function template specialization 'DB::DataPartStorageOnDiskBase::executeWriteOperation<(lambda at /home/kirillgarbar/ClickHouse/src/Storages/MergeTree/DataPartStorageOnDiskBase.cpp:621:35)>' requested here
  621 |             executeWriteOperation([&](auto & disk) { disk.removeSharedRecursive(to, true, {}); });
      |             ^
/home/kirillgarbar/ClickHouse/src/Disks/IDiskTransaction.h:112:18: note: 'removeSharedRecursive' declared here
  112 |     virtual void removeSharedRecursive(const std::string & path, bool /* keep_all_shared_data */, const NameSet & /* file_names_remove_metadata_only */, UInt64 /* batch_size */) = 0;
      |                  ^                     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
Template is instantiated both for disk and transaction types

Reply:
I see, method signatures is expected to be equal. How about putting setting to `DiskObjectStorageTransaction` in `createObjectStorageTransaction()`

Reply:
It is already [there](https://github.com/kirillgarbar/ClickHouse/blob/3405745f22ef7dc9a32ba732e50be18c602eb5d2/src/Disks/ObjectStorages/DiskObjectStorage.cpp#L45)
Or you mean something else?

Reply:
Missed it, then ok

Reply:
Or
How about create a new method in `IDiskTransaction` with restricted limit to delete and use it
 * in `DiskObjectstorage`: from `removeSharedRecursive` always call with limit `transaction->removeSharedRecursiveWithLimit`
 * In `IDiskTransaction`: from `removeSharedRecursive` always call with inf limit `removeSharedRecursiveWithLimit `